### DISCUSSION TITLE
When the subject in the certificate contains Chinese characters

### DISCUSSION THREAD

**Discussion by yw1573** [strongswan/strongswan#1825] [Q&A] [answered]
![image](https://github.com/strongswan/strongswan/assets/42687512/eba1ad5b-c82c-43d3-8378-d8db8c2b74b0)
![image](https://github.com/strongswan/strongswan/assets/42687512/8f8a9690-217a-4fd9-b6cc-25eb351919cf)
It cannot find the correct private key through the id in the configuration file
![image](https://github.com/strongswan/strongswan/assets/42687512/d9df78bf-58ed-4ae4-870e-ab222dc3cb36)
Although it looks like garbled characters, it is Chinese and correct
Please tell me how to adapt it to this situation

**Comment by yw1573** [comment]
![image](https://github.com/strongswan/strongswan/assets/42687512/3a524263-9f37-47cf-aa7c-9d68e41c2d0d)
![image](https://github.com/strongswan/strongswan/assets/42687512/b2711a49-b0e8-4d51-8cd1-c26acfa32011)
It seems that the id field configuration Chinese is not supported

**Reply by tobiasbrunner** [reply]
Might depend on the encoding of the file (UTF-8 might work as that will be the encoding in the RDN). But yeah, it's very much possible that the string-to-ASN.1 conversion has issues with Unicode characters. You could configure the identity as ASN.1 (`asn1dn:` prefix, can be extracted with [`pki --dn`](https://docs.strongswan.org/docs/5.9/pki/pkiDn.html)), or configure a SAN that doesn't contain any Unicode characters.

> Although it looks like garbled characters, it is Chinese and correct

Note that  strongSwan simply prints a `?` instead of any non-printable character (as defined by C's `isprint()`).

**Reply by yw1573** [reply]
You can look at my configuration and the logs I grabbed

**Reply by yw1573** [reply, marked as answer]
Verification passed
When the certificate contains special characters and non-ASCII characters, id configures the subject of the certificate, and the following methods can be used to obtain the subject code of the certificate：
```
pki --dn --in test.pem 
"asn1dn:#302a310b300906035504061302434e311b301906035504030c12313639313337333638392e36393733383232"
```
```
# in the configuration file
local {
auth = pubkey
id = "asn1dn:#302a310b300906035504061302434e311b301906035504030c12313639313337333638392e36393733383232"
certs = test.pem
cacerts = testCA.pem
}
```

**Comment by yw1573** [comment]
![image](https://github.com/strongswan/strongswan/assets/42687512/b37a5572-7dcb-4e34-b26a-2ce986a7d8e9)
```bash
shell# openssl x509 -in kele1_local.pem -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 6451353907117255353 (0x5987d032a20d6ab9)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=CN, ST=\xE5\x85\x94\xE5\x85\x94, L=\xE5\x85\x94\xE5\x85\x94, O=\xE5\x85\x94\xE5\x85\x94, OU=\xE5\x85\x94\xE5\x85\x94, CN=\xE5\x85\x94\xE5\x85\x94/emailAddress=tutu@tutu.com
        Validity
            Not Before: Aug  1 11:41:00 2023 GMT
            Not After : Jul 31 11:41:00 2024 GMT
        Subject: C=US, ST=\xE5\x8F\xAF\xE4\xB9\x90, L=\xE5\x8F\xAF\xE4\xB9\x90, O=\xE5\x8F\xAF\xE4\xB9\x90, OU=\xE5\x8F\xAF\xE4\xB9\x90, CN=\xE5\x8F\xAF\xE4\xB9\x90/emailAddress=kele@kele.com
shell# openssl x509 -in kele2_local.pem -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 5824459948477252217 (0x50d4a376bd7a0279)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=CN, ST=\xE5\x85\x94\xE5\x85\x94, L=\xE5\x85\x94\xE5\x85\x94, O=\xE5\x85\x94\xE5\x85\x94, OU=\xE5\x85\x94\xE5\x85\x94, CN=\xE5\x85\x94\xE5\x85\x94/emailAddress=tutu@tutu.com
        Validity
            Not Before: Aug  1 11:46:00 2023 GMT
            Not After : Jul 31 11:46:00 2024 GMT
        Subject: C=US, ST=\xE5\x8F\xAF\xE4\xB9\x90\xE5\x8F\xAF\xE4\xB9\x90, L=\xE5\x8F\xAF\xE4\xB9\x90\xE5\x8F\xAF\xE4\xB9\x90, O=\xE5\x8F\xAF\xE4\xB9\x90\xE5\x8F\xAF\xE4\xB9\x90, OU=\xE5\x8F\xAF\xE4\xB9\x90\xE5\x8F\xAF\xE4\xB9\x90, CN=\xE5\x8F\xAF\xE4\xB9\x90\xE5\x8F\xAF\xE4\xB9\x90/emailAddress=kele2@123.com
```
The above is my configuration and certificate
![image](https://github.com/strongswan/strongswan/assets/42687512/0cee1a18-ac70-4ec5-88a9-d2f12c5e6298)

**Reply by yw1573** [reply]
我的证书中的subj是这样的：
“C=US, ST=可乐可乐, L=可乐可乐, O=可乐可乐, OU=可乐可乐, CN=可乐可乐, E=kele2@123.com”
“C=US, ST=可乐, L=可乐, O=可乐, OU=可乐, CN=可乐, E=kele@kele.com”
I modified the config file
![image](https://github.com/strongswan/strongswan/assets/42687512/1af79a94-7a33-4d11-964b-653694d9e9bc)
```bash
shell:./swanctl -c test_fenzhi1
loading connection 'test_fenzhi1' failed: invalid value for: id, config discarded
loaded 0 (0 *) of 1 connections, 1 failed to load 
```

**Comment by yw1573** [comment]
pki --dn 
```bash
shell# pki --dn  --in kele2_local.pem 
"asn1dn:#30819e310b30090603550406130255533115301306035504080c0ce58fafe4b990e58fafe4b9903115301306035504070c0ce58fafe4b990e58fafe4b99031153013060355040a0c0ce58fafe4b990e58fafe4b99031153013060355040b0c0ce58fafe4b990e58fafe4b9903115301306035504030c0ce58fafe4b990e58fafe4b990311c301a06092a864886f70d010901160d6b656c6532403132332e636f6d"
shell# pki --dn  --in kele1_local.pem 
"asn1dn:#308180310b3009060355040613025553310f300d06035504080c06e58fafe4b990310f300d06035504070c06e58fafe4b990310f300d060355040a0c06e58fafe4b990310f300d060355040b0c06e58fafe4b990310f300d06035504030c06e58fafe4b990311c301a06092a864886f70d010901160d6b656c65406b656c652e636f6d"
How should I use the obtained asn1 and how to fill in the id field of the configuration file

**Reply by yw1573** [reply]
![image](https://github.com/strongswan/strongswan/assets/42687512/d06a4b96-e038-405c-aa91-21ce485a66ae)

**Reply by yw1573** [reply]
I am going to build an environment to test it, and then form a detailed test report

**Comment by yw1573** [comment]
![image](https://github.com/strongswan/strongswan/assets/42687512/8a057d4d-cdf4-4865-89ae-e6211a03ad74)
